File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cpp
#include<vector>
#include<bits/stdc++.h>
using namespace std;
int dx[4]={-1,1,0,0};
int dy[4]={0,0,-1,1};
int visited[104][104];

int solution(vector<vector<int> > maps)
{  
    visited[0][0]=1;
    queue<pair<int,int>> q; q.push({0,0});
    while(!q.empty()){
        int y = q.front().first;
        int x = q.front().second; 
        q.pop();
        if(y==maps.size()&&x==maps[0].size()) break;
        for(int i=0;i<4;i++){
            int ny = y+dy[i];int nx = x+dx[i];
            if(nx<0||ny<0||nx>=maps[0].size()||ny>=maps.size()||maps[ny][nx]==0||visited[ny][nx]) continue;
            q.push({ny,nx});
            visited[ny][nx]=visited[y][x]+1;
        }
    }
    int answer = visited[maps.size()-1][maps[0].size()-1]?visited[maps.size()-1][maps[0].size()-1]:-1;

    return answer;
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

dy = [-1,1,0,0]
dx = [0,0,-1,1]

def solution(maps):
    answer = -1
    queue = deque()
    n = len(maps)
    m = len(maps[0])
    queue.append((0,0))
    visited = [[0] * m for _ in range(n)]
    visited[0][0] = 1
    
    while queue:
        x, y = queue.popleft()
        if x == m - 1  and y == n - 1:
            return visited[n-1][m-1]
        for i in range(4):
            now_x = x + dx[i]
            now_y = y + dy[i]
            if 0 <= now_x < m and 0 <= now_y < n:
                if not visited[now_y][now_x] and maps[now_y][now_x] != 0:
                    visited[now_y][now_x] = visited[y][x] + 1
                    queue.append((now_x, now_y))

    return -1
